syncer/streamer_controller.go

```diff
@@ -0,0 +1,341 @@
+
+// Read remote binlog
+type remoteBinlogReader struct {
+	reader     *replication.BinlogSyncer
+	tctx       *tcontext.Context
+	EnableGTID bool
+}
+
+func (r *remoteBinlogReader) generateStreamer(pos mysql.Position) (streamer.Streamer, error) {
+	defer func() {
+		lastSlaveConnectionID := r.reader.LastConnectionID()
+		r.tctx.L().Info("last slave connection", zap.Uint32("connection ID", lastSlaveConnectionID))
+	}()
+
+	// FIXME: can enable GTID
+	if r.EnableGTID {
+		// NOTE: our (per-table based) checkpoint does not support GTID yet
+		return nil, terror.ErrSyncerUnitRemoteSteamerWithGTID.Generate()
+	}
+
+	// position's name may contain uuid, so need remove it
+	adjustedPos := binlog.AdjustPosition(pos)
+	streamer, err := r.reader.StartSync(adjustedPos)
+	return streamer, terror.ErrSyncerUnitRemoteSteamerStartSync.Delegate(err)
+}
+
+// StreamerController controls the streamer for read binlog, include:
+// 1. reopen streamer
+// 2. redirect binlog position or gtid
+// 3. transfor from local streamer to remote streamer
+type StreamerController struct {
+	sync.RWMutex
+
+	// the initital binlog type
+	initBinlogType BinlogType
+
+	// the current binlog type
+	currentBinlogType BinlogType
+
+	syncCfg        replication.BinlogSyncerConfig
+	localBinlogDir string
+	timezone       *time.Location
+
+	streamer         streamer.Streamer
+	streamerProducer StreamerProducer
+
+	// meetError means meeting error when get binlog event
+	// if binlogType is local and meetError is true, then need to create remote binlog stream
+	meetError bool
+
+	fromDB *UpStreamConn
+
+	uuidSuffix string
+}
+
+// NewStreamerController creates a new streamer controller
+func NewStreamerController(tctx *tcontext.Context, syncCfg replication.BinlogSyncerConfig, fromDB *UpStreamConn, binlogType BinlogType, localBinlogDir string, timezone *time.Location, beginPos mysql.Position) (*StreamerController, error) {
+	streamerController := &StreamerController{
+		initBinlogType:    binlogType,
+		currentBinlogType: binlogType,
+		syncCfg:           syncCfg,
+		localBinlogDir:    localBinlogDir,
+		timezone:          timezone,
+		fromDB:            fromDB,
+	}
+
+	err := streamerController.ResetReplicationSyncer(tctx, beginPos)
+	if err != nil {
+		streamerController.Close(tctx)
+		return nil, err
+	}
+
+	return streamerController, nil
+}
+
+// ResetReplicationSyncer reset the replication
+func (c *StreamerController) ResetReplicationSyncer(tctx *tcontext.Context, pos mysql.Position) (err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	uuidSameWithUpstream := true
+
+	// close old streamerProducer
+	if c.streamerProducer != nil {
+		switch t := c.streamerProducer.(type) {
+		case *remoteBinlogReader:
+			c.closeBinlogSyncer(tctx, t.reader)
+		case *localBinlogReader:
+			// check the uuid before close
+			uuidSameWithUpstream, err = c.checkUUIDSameWithUpstream(pos, t.reader.GetUUIDs())
+			if err != nil {
+				return err
+			}
+			// TODO: close old local reader before creating a new one
```

Comment:
Do we need to implement this TODO now?

Reply:
just use `t.reader.Close()` is ok, update in f99706a